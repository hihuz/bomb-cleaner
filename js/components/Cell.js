import React from 'react';

const Cell = ({ handleLeftClick, handleRightClick, index, opened, flagged, value }) => {
  const colors = ['#000', '#0100fe', '#017f01', '#fe0000', '#010080', '#810102', '#008081', '#000', '#808080'];
  const styles = {
    color: value !== 'X' && value !== ' ' && opened ? colors[value] : colors[0]
  };
  function leftClick(e) {
    e.preventDefault();
    if (handleLeftClick) { handleLeftClick(index, value); }
  }
  function rightClick(e) {
    e.preventDefault();
    if (handleRightClick) { handleRightClick(index); }
  }
  return (
    <div
      className={`cell${opened ? ' opened' : ''}${value === 'X' && opened ? ' bomb' : ''}`}
      onClick={leftClick}
      onContextMenu={rightClick}
      style={styles}
    >
      {opened && value !== 'X' ? value : ' '}
      {opened && value === 'X' ? <i className="icon-bomb" /> : ' '}
      {flagged && !opened ? <i className="icon-flag" /> : ''}
    </div>
  );
};

export default Cell;
