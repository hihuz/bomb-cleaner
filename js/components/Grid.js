import React from 'react';
import Cell from './Cell';
import { connect } from 'react-redux';
import { toggleFlag, openBomb, openCell } from '../actions/actionCreators';

const Grid = ({ grid, status, dispatch }) => {
  function handleLeftClick(index, value) {
    switch (value) {
      case "X":
        return dispatch(openBombCell(index));
      default:
        return dispatch(openCell(index));
    }
  }
  function handleRightClick(index) {
      dispatch(toggleFlag(index));
  }
  const styles = {
    width: grid.width * 24 + "px",
    height: grid.height * 24 + "px"
  }

  return (
    <div className="grid" style={styles}>
      {grid.cells.map((cell) => {
        return (
          <Cell
            handleLeftClick={ cell.opened ||
                              cell.flagged ||
                              status == 'lost' ||
                              status == 'won' ? undefined : handleLeftClick }
            handleRightClick={ cell.opened ||
                               status == 'lost' ||
                               status == 'won' ? undefined : handleRightClick }
            key={cell.index}
            {...cell}
          />
        );
      })}
    </div>
  );
}

export default connect()(Grid);
