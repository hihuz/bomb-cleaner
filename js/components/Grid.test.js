import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Plain } from './Grid';
import Cell from './Cell';

function makeTestGrid(gridSize, gameStatus, cellsFlagged, cellsOpened) {
  const size = gridSize || 20;
  const status = gameStatus || 'running';
  const flagged = cellsFlagged || false;
  const opened = cellsOpened || false;
  const testGridProp = {
    width: 1,
    height: 1,
    cells: new Array(size).fill(undefined).map((cell, i) => ({
      index: i,
      flagged,
      opened,
      value: 2
    }))
  };
  return shallow(<Plain grid={testGridProp} status={status} />);
}

test('Snapshot', () => {
  const component = makeTestGrid();
  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});

test('Should render a number of cells based off grid.cells prop', () => {
  const gridSize = 80;
  const component = makeTestGrid(gridSize);
  expect(component.find(Cell).length).toEqual(gridSize);
});

test('Should pass cell props correctly to child Cell component', () => {
  const gridSize = 1;
  const component = makeTestGrid(gridSize);
  const child = component.find(Cell);
  const childProps = child.props();
  expect(child.length).toEqual(gridSize);
  expect(childProps.index).toEqual(0);
  expect(childProps.flagged).toEqual(false);
  expect(childProps.opened).toEqual(false);
  expect(childProps.value).toEqual(2);
  expect(childProps.handleLeftClick).toBeDefined();
  expect(childProps.handleRightClick).toBeDefined();
});

test('Should NOT pass handleLeftClick/handleRightClick to an opened cell', () => {
  const component = makeTestGrid(1, 'running', false, true);
  const child = component.find(Cell);
  const childProps = child.props();
  expect(childProps.handleLeftClick).toBeUndefined();
  expect(childProps.handleRightClick).toBeUndefined();
});

test('Should NOT pass handleLeftClick to a flagged cell, but should pass handleRightClick', () => {
  const component = makeTestGrid(1, 'running', true, false);
  const child = component.find(Cell);
  const childProps = child.props();
  expect(childProps.handleLeftClick).toBeUndefined();
  expect(childProps.handleRightClick).toBeDefined();
});

test('Should NOT pass handleLeftClick/handleRightClick if the game is won', () => {
  const component = makeTestGrid(1, 'won', false, false);
  const child = component.find(Cell);
  const childProps = child.props();
  expect(childProps.handleLeftClick).toBeUndefined();
  expect(childProps.handleRightClick).toBeUndefined();
});

test('Should NOT pass handleLeftClick/handleRightClick if the game is lost', () => {
  const component = makeTestGrid(1, 'lost', false, false);
  const child = component.find(Cell);
  const childProps = child.props();
  expect(childProps.handleLeftClick).toBeUndefined();
  expect(childProps.handleRightClick).toBeUndefined();
});
