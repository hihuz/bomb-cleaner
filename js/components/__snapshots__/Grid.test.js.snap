exports[`test Snapshot 1`] = `
<div
  className="grid"
  style={
    Object {
      "height": "24px",
      "width": "24px",
    }
  }>
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={0}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={1}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={2}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={3}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={4}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={5}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={6}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={7}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={8}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={9}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={10}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={11}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={12}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={13}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={14}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={15}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={16}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={17}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={18}
    opened={false}
    value={2} />
  <Cell
    flagged={false}
    handleLeftClick={[Function]}
    handleRightClick={[Function]}
    index={19}
    opened={false}
    value={2} />
</div>
`;
